kernel: bump 5.4 to 5.4.50

graysky2 opened the pull request:
updated `include/kernel-version.mk` and `make target/linux/refresh V=s`
    
Run tested: R7800
Build tested: ipq806x

mans0n commented:
Could you explain for the ledtrig-libata.patch change? It seems unnecessary to me.

graysky2 commented:
@mans0n - Without that change, the patch did not apply.

graysky2 commented:
@mans0n - I needed to modify it or else it does not apply to 5.4.49:
```
Applying /incoming/openwrt/target/linux/generic/pending-5.4/834-ledtrig-libata.patch using plaintext: 
patching file drivers/ata/Kconfig
patching file drivers/ata/libata-core.c
Hunk #1 succeeded at 714 (offset -1 lines).
Hunk #2 succeeded at 5128 (offset -1 lines).
Hunk #3 succeeded at 6065 (offset -1 lines).
Hunk #4 succeeded at 6103 (offset -1 lines).
Hunk #5 succeeded at 6572 (offset -1 lines).
patching file include/linux/libata.h
Hunk #1 FAILED at 22.
Hunk #2 succeeded at 879 (offset 3 lines).
1 out of 2 hunks FAILED -- saving rejects to file include/linux/libata.h.rej
Patch failed!  Please fix /incoming/openwrt/target/linux/generic/pending-5.4/834-ledtrig-libata.patch!
make[4]: *** [Makefile:34: /incoming/openwrt/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-ipq806x_generic/linux-5.4.49/.prepared_e2cf3ec102295cd68c95307fda0b470e] Error 1
make[4]: Leaving directory '/incoming/openwrt/target/linux/ipq806x'
make[3]: *** [Makefile:13: compile] Error 2
make[3]: Leaving directory '/incoming/openwrt/target/linux'
time: target/linux/compile#7.70#2.84#16.90
make[2]: *** [target/Makefile:25: target/linux/compile] Error 2
make[2]: Leaving directory '/incoming/openwrt'
make[1]: *** [target/Makefile:18: /incoming/openwrt/staging_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/stamp/.target_compile] Error 2
```

mans0n commented on target/linux/generic/pending-5.4/834-ledtrig-libata.patch:
```diff
@@ -122,28 +122,34 @@ Signed-off-by: Daniel Golle <daniel@makrotopia.org>
 +#ifdef CONFIG_ATA_LEDS
 +#include <linux/leds.h>
 +#endif
+ #include <linux/async.h>
```
Okay, now I see that this line is required, but other lines seems trivial or duplicate.
I would suggest something like this:
```diff
--- a/target/linux/generic/pending-5.4/834-ledtrig-libata.patch
+++ b/target/linux/generic/pending-5.4/834-ledtrig-libata.patch
@@ -125,9 +125,9 @@ Signed-off-by: Daniel Golle <daniel@makrotopia.org>
 --- a/include/linux/libata.h
 +++ b/include/linux/libata.h
 @@ -22,6 +22,9 @@
- #include <linux/acpi.h>
  #include <linux/cdrom.h>
  #include <linux/sched.h>
+ #include <linux/async.h>
 +#ifdef CONFIG_ATA_LEDS
 +#include <linux/leds.h>
 +#endif
```

graysky2 commented:
@mans0n - Agreed.  The new rebased commit patches cleanly and OW is running just fine on my R7800 after compiling with it.

mans0n commented:
If you are manually editing patches, you can try an automated patch refresh command `make target/linux/refresh V=s`.

And there's also a [script](https://git.openwrt.org/?p=maintainer-tools.git;a=blob;f=update_kernel.sh;h=0cbdb1d28f0b9466d528ddaa2a7b8ec7aa95a117;hb=HEAD) for kernel update (though I've never tried it).

graysky2 commented:
I will check that out for future work.  I think the PR has been edited per guidance.  Let me know if suitable.

mans0n commented:
There might be some remaining patches that should be refreshed in other targets, so AFAIK we should use the script I mentioned above for kernel version update.
If you already did a refresh manually for all targets then it might be okay.

graysky2 commented:
I did it, 18 files now affected.  Most seem cosmetic but I like to observe best practices.  Built find/booted into it fine.  Have a look and let me know if anything else is needed.

ynezz commented:
Every kernel bump needs to refresh/rebase all patches in all targets with `maintainer-tools/update_kernel.sh -v -u 5.4` command. You can find the tools at the Git repo https://git.openwrt.org/maintainer-tools.git

graysky2 commented:
@ynezz - Thanks for the tip.  What are the correct order steps?

1) Update `include/kernel-version.mk` with new version
2) Run `update_kernel.sh -v -u 5.4`
3) Compile and flash?

ynezz commented:
Just `2.`, that `-u` will update `include/kernel-version.mk` as well.

graysky2 commented:
@ynezz - I tried running it against master:
```
Refreshing Kernel 5.4 to release 5.4.50 ...
refreshing apm821xx ...
#
# configuration written to .config
#
time: target/linux/prereq#0.09#0.02#0.11
make[1]: Entering directory '/scratch/union'
make[2]: Entering directory '/scratch/union/target/linux'
make[3]: Entering directory '/scratch/union/target/linux/apm821xx'
mkdir -p /scratch/union/dl
SHELL= flock /scratch/union/tmp/.linux-5.4.50.tar.xz.flock -c '  	/scratch/union/scripts/download.pl "/scratch/union/dl" "linux-5.4.50.tar.xz" "skip" "" "@KERNEL/linux/kernel/v5.x"    '
+ curl -f --connect-timeout 20 --retry 5 --location --insecure https://sources.cdn.openwrt.org/linux-5.4.50.tar.xz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404 
Download failed.
+ curl -f --connect-timeout 20 --retry 5 --location --insecure https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.4.50.tar.xz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  104M  100  104M    0     0  22.0M      0  0:00:04  0:00:04 --:--:-- 23.4M
rm -rf /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand
mkdir -p /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand
xzcat /scratch/union/dl/linux-5.4.50.tar.xz | tar -C /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand -xf -
rm -rf /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50/patches; mkdir -p /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50/patches
cp -fpR "/scratch/union/target/linux/generic/files"/. "/scratch/union/target/linux/apm821xx/files"/. /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50/
find /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50/ -name \*.rej -or -name \*.orig | xargs -r rm -f
if [ -d /scratch/union/target/linux/generic/patches ]; then echo "generic patches directory is present. please move your patches to the pending directory" ; exit 1; fi
touch /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50/.quilt_used
touch /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50/.prepared
if [ -s "/scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50/patches/series" ]; then (cd "/scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50"; if quilt --quiltrc=- next >/dev/null 2>&1; then quilt --quiltrc=- push -a; else quilt --quiltrc=- top >/dev/null 2>&1; fi ); fi
Applying patch generic-backport/010-Kbuild-don-t-hardcode-path-to-awk-in-scripts-ld-vers.patch
patching file scripts/ld-version.sh

Applying patch generic-backport/011-kbuild-export-SUBARCH.patch
patching file Makefile

Applying patch generic-backport/020-backport_netfilter_rtcache.patch
patching file include/net/netfilter/nf_conntrack_extend.h
patching file include/net/netfilter/nf_conntrack_rtcache.h
patching file net/netfilter/Kconfig
patching file net/netfilter/Makefile
patching file net/netfilter/nf_conntrack_rtcache.c

Applying patch generic-backport/030-modpost-add-a-helper-to-get-data-pointed-by-a-symbol.patch
patching file scripts/mod/modpost.c

Applying patch generic-backport/031-modpost-refactor-namespace_from_kstrtabns-to-not-har.patch
patching file scripts/mod/modpost.c
patching file scripts/mod/modpost.h

Applying patch generic-backport/040-ARM-8918-2-only-build-return_address-if-needed.patch
patching file arch/arm/kernel/Makefile
patching file arch/arm/kernel/return_address.c

Applying patch generic-backport/300-MIPS-Exclude-more-dsemul-code-when-CONFIG_MIPS_FP_SU.patch
patching file arch/mips/include/asm/processor.h
patching file arch/mips/kernel/process.c
patching file arch/mips/kernel/vdso.c

Applying patch generic-backport/343-netfilter-nft_flow_offload-handle-netdevice-events-f.patch
patching file net/netfilter/nf_flow_table_core.c
patching file net/netfilter/nft_flow_offload.c

Applying patch generic-backport/370-netfilter-nf_flow_table-fix-offloaded-connection-tim.patch
patching file net/netfilter/nf_conntrack_core.c
patching file net/netfilter/nf_flow_table_core.c

Applying patch generic-backport/371-netfilter-nf_flow_table-fix-up-ct-state-of-flows-aft.patch
patching file net/netfilter/nf_flow_table_core.c

Applying patch generic-backport/393-v5.5-sch_cake-drop-unused-variable-tin_quantum_prio.patch
patching file net/sched/sch_cake.c
Hunk #2 succeeded at 1942 (offset 24 lines).
Hunk #3 succeeded at 2264 (offset 24 lines).
Hunk #4 succeeded at 2275 (offset 24 lines).
Hunk #5 succeeded at 2288 (offset 24 lines).
Hunk #6 succeeded at 2364 (offset 24 lines).
Hunk #7 succeeded at 2380 (offset 24 lines).
Hunk #8 succeeded at 2426 (offset 24 lines).
Hunk #9 succeeded at 2461 (offset 24 lines).

Applying patch generic-backport/395-v5.8-net-sch_cake-Take-advantage-of-skb-hash-where-appropriate.patch
patching file net/sched/sch_cake.c

Applying patch generic-backport/396-5.8-sch_cake-don-t-try-to-reallocate-or-unshare-skb-unco.patch
patching file net/sched/sch_cake.c
Hunk #1 FAILED at 1553.
1 out of 1 hunk FAILED -- rejects in file net/sched/sch_cake.c
Patch generic-backport/396-5.8-sch_cake-don-t-try-to-reallocate-or-unshare-skb-unco.patch can be reverse-applied
make[3]: *** [Makefile:27: /scratch/union/build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50/.quilt_checked] Error 1
make[3]: Leaving directory '/scratch/union/target/linux/apm821xx'
make[2]: *** [Makefile:13: refresh] Error 2
make[2]: Leaving directory '/scratch/union/target/linux'
time: target/linux/refresh#6.92#2.17#11.99
make[1]: *** [target/Makefile:25: target/linux/refresh] Error 2
make[1]: Leaving directory '/scratch/union'
make: *** [/scratch/union/include/toplevel.mk:224: target/linux/refresh] Error 2
```

ynezz commented:
>  Is it expected to have to manually fix patches that fail?

Manually or you can learn that script to automagically remove/skip those `can be reverse-applied` upstreamed patches.

graysky2 commented:
I'm not sure how to proceed... if `series` lists `396-5.8-sch_cake-don-t-try-to-reallocate-or-unshare-skb-unco.patch` but presumably 5.4.50 includes that patch, what is the correct action to take?  Remove that line from `series`?  `git rm target/linux/generic/backport-5.4/396-5.8-sch_cake-don-t-try-to-reallocate-or-unshare-skb-unco.patch`?

graysky2 commented:
1) It seems that the following patches have been upstreamed as they end with `xxx can be reverse-applied` ... so I commented their respective lines in `build_dir/target-powerpc_464fp_musl/linux-apm821xx_nand/linux-5.4.50/patches/series` and re-ran the script.

```
#generic-backport/396-5.8-sch_cake-don-t-try-to-reallocate-or-unshare-skb-unco.patch
#generic-backport/397-5.8-sch_cake-don-t-call-diffserv-parsing-code-when-it-is.patch
#generic-backport/398-5.8-sch_cake-fix-a-few-style-nits.patch
#generic-backport/812-v5.8-i2c-pxa-fix-i2c_pxa_scream_blue_murder-debug-output.patch
#generic-backport/819-v5.8-i2c-pxa-clear-all-master-action-bits-in-i2c_pxa_stop.patch
```

2) I manually edited `target/linux/generic/pending-5.4/834-ledtrig-libata.patch` per the discussion above to allow it to apply.

The script has been running for 20 minutes now looping through the targets.  I will report back when it finishes or when I get stuck.

graysky2 commented:
It failed here:
```
...
Applying patch platform/701-net-0232-enetc-Handle-USXGMII-protocol.patch
patching file drivers/net/ethernet/freescale/enetc/enetc_pf.c

Applying patch platform/701-net-0233-enetc-Enable-live-mac-addr-change-for-PF.patch
patching file drivers/net/ethernet/freescale/enetc/enetc_pf.c

Applying patch platform/701-net-0234-enetc-WA-for-MDIO-register-access-issue.patch
patching file drivers/net/ethernet/freescale/enetc/enetc.c
Hunk #4 FAILED at 259.
Hunk #6 FAILED at 300.
2 out of 15 hunks FAILED -- rejects in file drivers/net/ethernet/freescale/enetc/enetc.c
patching file drivers/net/ethernet/freescale/enetc/enetc_hw.h
patching file drivers/net/ethernet/freescale/enetc/enetc_mdio.c
patching file drivers/net/ethernet/freescale/enetc/enetc_pf.c
Patch platform/701-net-0234-enetc-WA-for-MDIO-register-access-issue.patch does not apply (enforce with -f)
make[3]: *** [Makefile:23: /scratch/openwrt/build_dir/target-aarch64_generic_musl/linux-layerscape_armv8_64b/linux-5.4.50/.quilt_checked] Error 1
make[3]: Leaving directory '/scratch/openwrt/target/linux/layerscape'
make[2]: *** [Makefile:13: refresh] Error 2
make[2]: Leaving directory '/scratch/openwrt/target/linux'
time: target/linux/refresh#11.07#2.84#12.39
make[1]: *** [target/Makefile:25: target/linux/refresh] Error 2
make[1]: Leaving directory '/scratch/openwrt'
make: *** [/scratch/openwrt/include/toplevel.mk:224: target/linux/refresh] Error 2
```

I took a crack at manually inspecting the failed hunks.  I think I got the ones around `drivers/net/ethernet/freescale/enetc/enetc.c` OK but the last ones around `drivers/net/ethernet/freescale/enetc/enetc_mdio.c` seems beyond my limited skills.

[Here](https://gist.github.com/graysky2/7fcfe06faa75c490de692ab69bab7c5a) is the my attempt to partially fix it.

graysky2 commented:
Rebased and pushed for feedback.

Commit:
kernel: bump 5.4 to 5.4.50
update via update_kernel.sh -v -u 5.4

Resolved merge conflict in the following patches:
 701-net-0234-enetc-WA-for-MDIO-register-access-issue.patch
 834-ledtrig-libata.patch

Removed upstreamed patches:
 396-5.8-sch_cake-don-t-try-to-reallocate-or-unshare-skb-unco.patch
 397-5.8-sch_cake-don-t-call-diffserv-parsing-code-when-it-is.patch
 398-5.8-sch_cake-fix-a-few-style-nits.patch
 812-v5.8-i2c-pxa-fix-i2c_pxa_scream_blue_murder-debug-output.patch
 819-v5.8-i2c-pxa-clear-all-master-action-bits-in-i2c_pxa_stop.patch

All others refreshed

build tested: ipq806x (netgear R7800)

Signed-off-by: John Audia graysky@archlinux.us

ynezz commented:
> but it really needs someone more knowledgeable than I to review it for sanity.

@ldir-EDB0 has it as well https://git.openwrt.org/?p=openwrt/staging/ldir.git;a=commit;h=416a70d47c90726d3f189febebdd97b1f9b8957c so you can compare

BTW your commit description is incomplete, you did deleted few patches so it should be described why, you've also solved merge conflict in the target/linux/generic/pending-5.4/834-ledtrig-libata.patch, for the commit description inspiration just take a look at the output of `git log include/kernel-version.mk`.

graysky2 commented:
@ynezz - I see.  Since Kevin already covered this code, shall we just close this and use his?  In the future, I know to use that script and provide details per your suggestion.

hauke commented:
A similar version was applied to master.
